File: src/main/java/com/polygloat/dtos/request/CreateApiKeyDTO.java
package com.polygloat.dtos.request;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.polygloat.constants.ApiScope;
import lombok.*;

import javax.validation.Valid;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.Set;
import java.util.stream.Collectors;

@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CreateApiKeyDTO {
    @NotNull
    @Getter
    @Setter
    private Long repositoryId;

    @JsonIgnore
    @NotEmpty
    @Getter
    @Setter
    private Set<ApiScope> scopes;

    @JsonSetter("scopes")
    public void jsonSetScopes(Set<String> scopes) {
        this.scopes = scopes.stream().map(ApiScope::fromValue).collect(Collectors.toSet());
    }

    @JsonGetter("scopes")
    public Set<String> jsonGetScopes() {
        return this.scopes.stream().map(ApiScope::getValue).collect(Collectors.toSet());
    }
}

File: backend/data/src/main/kotlin/io/tolgee/dtos/request/apiKey/CreateApiKeyDto.kt
package io.tolgee.dtos.request.apiKey

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonSetter
import io.swagger.v3.oas.annotations.media.Schema
import io.tolgee.model.enums.Scope
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull
import org.hibernate.validator.constraints.Length
import java.util.stream.Collectors

data class CreateApiKeyDto(
  @field:NotNull
  @field:Min(1)
  var projectId: Long = 0,
  @field:NotEmpty
  @JsonIgnore
  @Schema(
    example = """
    ["screenshots.upload", "screenshots.delete", "translations.edit", "screenshots.view", "translations.view", "keys.edit"]
    """,
  )
  var scopes: Set<Scope> = setOf(),
  @Schema(description = "Description of the project API key")
  @field:Length(max = 250, min = 1)
  var description: String? = null,
  @Schema(
    description =
      "Expiration date in epoch format (milliseconds)." +
        " When null key never expires.",
    example = "1661172869000",
  )
  val expiresAt: Long? = null,
) {
  @Suppress("unused")
  @JsonSetter("scopes")
  fun jsonSetScopes(scopes: Set<String>) {
    this.scopes = scopes.map { value -> Scope.fromValue(value) }.toSet()
  }

  @Suppress("unused")
  @JsonGetter("scopes")
  fun jsonGetScopes(): Set<String> {
    return scopes.stream().map { obj: Scope -> obj.value }.collect(Collectors.toSet())
  }
}
